dnf/transaction_sr.py

```diff
@@ -0,0 +1,359 @@
+                    libdnf.transaction.compsPackageTypeToString(group.getPackageTypes())
+
+            groups.append(group_data)
+
+        elif tsi.is_environment():
+            env = tsi.get_environment()
+
+            env_data = {
+                "action": tsi.action_name,
+                "id": env.getEnvironmentId(),
+            }
+
+            if tsi.action == libdnf.transaction.TransactionItemAction_INSTALL:
+                env_data["package_types"] = \
+                    libdnf.transaction.compsPackageTypeToString(env.getPackageTypes())
+
+            environments.append(env_data)
+
+    if rpms:
+        data["rpms"] = rpms
+
+    if groups:
+        data["groups"] = groups
+
+    if environments:
+        data["environments"] = environments
+
+    with open(filename, "w") as f:
+        json.dump(data, f, indent=4, sort_keys=True)
+        f.write("\n")
+
+    print(_("Transaction saved to {}.").format(filename))
+
+
+class TransactionReplay(object):
+    def __init__(self, base, fn):
+        self._base = base
+        self._filename = fn
+
+    def _verify_toplevel_json(self, replay_data):
+        fn = self._filename
+
+        if "version" not in replay_data:
+            raise TransactionFileError(fn, _('Missing key "{key}".'.format(key="version")))
+
+        if not isinstance(replay_data["version"], str):
+            raise TransactionFileError(fn, _('Unexpected type of "version", string expected.'))
+
+        check_version(replay_data["version"], fn)
+
+    def _replay_pkg_action(self, pkg_data):
+        try:
+            action = pkg_data["action"]
+            nevra = pkg_data["nevra"]
+            reason = pkg_data["reason"]
+        except KeyError as e:
+            raise TransactionError(
+                _('Missing object key "{key}" in an rpm.').format(key=e.args[0])
+            )
+
+        solution = dnf.subject.Subject(nevra).get_best_solution(
+            self._base.sack,
+            with_nevra=True,
+            with_provides=False,
+            with_filenames=False,
+            forms=[hawkey.FORM_NEVRA]
+        )
+
+        if solution["nevra"] is None:
+            raise TransactionError(_('Cannot find rpm nevra "{nevra}".').format(nevra=nevra))
+
+        query = solution["query"]
+        if not query:
+            raise TransactionError(_('Rpm nevra "{nevra}" was not found.').format(nevra=nevra))
+
+        # validate the reasons before running the transaction and cache them
+        try:
+            self._pkg_reason_cache[nevra] = \
+                libdnf.transaction.StringToTransactionItemReason(reason)
+        except IndexError as e:
+            raise TransactionError(
+                _('Unexpected value of package reason "{reason}" for rpm nevra "{nevra}".')
+                    .format(reason=reason, nevra=nevra)
+            )
+
+        if action in ("Install", "Upgrade", "Downgrade", "Reinstall"):
+            sltr = dnf.selector.Selector(self._base.sack).set(pkg=query.available())
```

Comment:
In case of "Install", "Upgrade", "Downgrade", the transaction could result in in reinstall. Try to run `dnf install acpi`, store the transaction, then replay it. It results in reinstall.
```
        if action in ("Install", "Upgrade", "Downgrade", "Reinstall"):
            sltr = dnf.selector.Selector(self._base.sack).set(pkg=query)
        elif action == "Reinstall":
            available = query.available()
            if not available:
                raise ...
            sltr = dnf.selector.Selector(self._base.sack).set(pkg=available)
```

Reply:
It does result in a reinstall if you re-run the transaction on a system where the transaction was already performed.

In case of "Upgrade" and "Downgrade" the replay would fail on the version check of the installed package (for the "Upgraded" and "Downgraded" actions of the appropriate package NA). I guess to keep the same level of strictness for "Install" as well, I should check that the package NA is not installed at all.

Understandably this is not very useful and we should have a discussion about the strictness levels and what should be the default (maybe we don't want the most strict mode to be the default, but we should at least be consistent with the strictness of the various actions).

It is not at all clear what your code snippet is supposed to do, seems to me it would result in not actually reinstalling the package (i.e. result in a noop) in case the target NEVRA is already on the system. That sounds like a good idea... but the strictness question remains.